modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/OrphanDetector.java

```diff
@@ -0,0 +1,179 @@
+            //LockManager lockManager,
+            HybridClock clock) {
+        this.topologyService = topologyService;
+        this.replicaService = replicaService;
+        this.placementDriver = placementDriver;
+        //this.lockManager = lockManager;
+        this.clock = clock;
+    }
+
+    /**
+     * Starts the detector.
+     *
+     * @param txLocalStateStorage Local transaction state storage.
+     */
+    public void start(Function<UUID, TxStateMeta> txLocalStateStorage) {
+        this.txLocalStateStorage = txLocalStateStorage;
+        // TODO: IGNITE-20773 Subscribe to lock conflicts here.
+    }
+
+    /**
+     * Stops the detector.
+     */
+    public void stop() {
+        busyLock.block();
+        // TODO: IGNITE-20773 Unsubscribe from lock conflicts here.
+    }
+
+    /**
+     * Sends {@link TxRecoveryMessage} if the transaction is orphaned.
+     * TODO: IGNITE-20773 Invoke the method when the lock conflict is noted.
+     *
+     * @param txId Transaction id that holds a lock.
+     * @return Future to complete.
+     */
+    private CompletableFuture<Void> handleLockHolder(UUID txId) {
+        if (busyLock.enterBusy()) {
+            try {
+                return handleLockHolderInternal(txId);
+            } finally {
+                busyLock.leaveBusy();
+            }
+        }
+
+        return failedFuture(new NodeStoppingException());
+    }
+
+    /**
+     * Sends {@link TxRecoveryMessage} if the transaction is orphaned.
+     *
+     * @param txId Transaction id that holds a lock.
+     * @return Future to complete.
+     */
+    private CompletableFuture<Void> handleLockHolderInternal(UUID txId) {
+        TxStateMeta txState = txLocalStateStorage.apply(txId);
+
+        assert txState != null : "The transaction is undefined in the local node [txId=" + txId + "].";
+
+        if (topologyService.getById(txState.txCoordinatorId()) == null) {
+            LOG.info(
+                    "Conflict was found, and the coordinator of the transaction that holds a lock is not available "
+                            + "[txId={}, txCrd={}].",
+                    txId,
+                    txState.txCoordinatorId()
+            );
+
+            return placementDriver.awaitPrimaryReplica(
+                    txState.commitPartitionId(),
+                    clock.now(),
+                    AWAIT_PRIMARY_REPLICA_TIMEOUT_SEC,
+                    SECONDS
+            ).thenCompose(replicaMeta -> {
+                ClusterNode commitPartPrimaryNode = topologyService.getByConsistentId(replicaMeta.getLeaseholder());
+
+                if (commitPartPrimaryNode == null) {
+                    LOG.warn(
+                            "The primary replica of the commit partition is not available [commitPartGrp={}, tx={}]",
+                            txState.commitPartitionId(),
+                            txId
+                    );
+
+                    return completedFuture(null);
```

Comment:
Failed future should be returned from here, it will allow us to mark this tx as abandoned (or can we do it right here?)